### DISCUSSION TITLE
Node coloring

### DISCUSSION THREAD

**Discussion by DarkFenX** [networkx/networkx#4568] [Q&A] [answered]
Hi, I have two sets of data I am trying to visualize on my graph, both are numerical. I want to show node name and two numbers (which could be embedded into node name if there are no other options), and color node according to those numbers. Solutions I could think of would include sharp color split (left/right half of node) or gradient. Unfortunately, it seems that matplotlib (by proxy of networkx's `draw` call) does not accept matplotlib colormaps as a color argument.

Do I have any way to achieve what I want, or do I have to code something with lower-level tools, like PIL or gaphas?

**Comment by rossbar** [comment, marked as answer]
Matplotlib [does support bicolor markers](https://matplotlib.org/3.3.3/gallery/lines_bars_and_markers/marker_reference.html#marker-fill-styles), but unfortunately the relevant keywords won't be passed through from the `nx_pylab` functions. However, you should be able to accomplish your goal using matplotlib directly.

For example, let's say you have a graph with `val1` ranging from [0, 20), and `val2` ranging from [-100, 100). You could plot these values on a per-node basis with independent colormaps doing something like:

```python
import networkx as nx
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm
from matplotlib.colors import Normalize
rng = np.random.default_rng()

# Generate a path graph with n nodes and two node attributes:
# val1, which ranges from [0, 20), and val2 ranging from [-100, 100)
n = 10
G = nx.path_graph(n)
for name, (minval, maxval) in zip(('val1', 'val2'), ((0, 20), (-100, 100))):
    nx.set_node_attributes(
        G,
        {k: v for k, v in zip(range(n), rng.integers(minval, maxval, n))},
        name=name,
    )

# Get layout
pos = nx.circular_layout(G)

# Set up colormaps and normalizations for the two different node attributes
val1_cmap = cm.get_cmap('autumn')
val2_cmap = cm.get_cmap('winter')
val1_norm = Normalize(vmin=0, vmax=20)
val2_norm = Normalize(vmin=-100, vmax=100)

# Default values for filled marker style
filled_marker_style = dict(
    marker='o',
    markersize=25,
    markeredgecolor='k'
)

# Plot each node independently with its own marker style determined by the
# node values
for node in G.nodes():
    x, y = pos[node]
    filled_marker_style['markerfacecolor'] = val1_cmap(val1_norm(G.nodes[node]['val1']))
    filled_marker_style['markerfacecoloralt'] = val2_cmap(val2_norm(G.nodes[node]['val2']))
    plt.plot(x, y, fillstyle='left', **filled_marker_style)

plt.margins(0.10)
plt.axis('off')
plt.show()
```
Which gives something like:
![Figure_1](https://user-images.githubusercontent.com/1268991/105783026-643e4600-5f2a-11eb-83d2-9378e690e33c.png)

This is just a basic example to illustrate the main concepts. Hope it's helpful!

**Reply by dschult** [reply]
Nice @rossbar!

Then to draw edges use `nx.draw_networkx_edges(G, pos, node_size=25)`

**Reply by DarkFenX** [reply]
Thanks a ton! I do not mind to use MPL directly / make locally adjusted version of networkx to handle that, just being able to do it without lower-level tools is great.
